### ISSUE TITLE
Create bot that automatically reports Alexa rankings

### ISSUE THREAD

**Issue by ericcornelissen** [simple-icons/simple-icons#982]
As per recent discussion on #857. We're thinking about creating a bot for GitHub, most likely using [probot](https://github.com/probot/probot), that automatically reports the [Alexa ranking](https://www.alexa.com/siteinfo) of a website in Issues and Pull Requests(?), possible approaches:

<details>
 <summary>Report the rank of any link found in the original post of a Issue/Pull Request</summary>
 <b>Pros</b>: Easy to implement, easy for contributors<br>
 <b>Cons</b>: Probably very spammy (e.g. wikipedia, wikimedia, dropbox, seekvectorlogo)
</details>
<br>
<details>
 <summary>Respond to a link at a specific point in the original post of an Issue/Pull Request. This requires a clear <a href="https://blog.github.com/2016-02-17-issue-and-pull-request-templates/">template</a>.</summary>
 <b>Pros</b>: Relatively easy for contributors<br>
 <b>Cons</b>: First time contributors may mess up
</details>
<br>
<details>
 <summary>Respond to someone commenting (e.g.) `AlexaBot! google.com`</summary>
 <b>Pros</b>: Easy to implement; easy to use<br>
 <b>Cons</b>: Feedback is not instant
</details>
<br>
<details>
 <summary>Report the rank of links found in the div</summary>
 <b>Pros</b>: Requires no additional work from contributors<br>
 <b>Cons</b>: Hard to implement, Does not work for issues, Not always to the actual site (e.g. wikimedia links)
</details>
<br>

Comments and suggestions are welcome 😃

**Comment by birjolaxew** [comment]
I'd suggest reporting rank on all URLs, except a blacklist that we manually maintain (and only in the issue body, not in any comments). It isn't *that* big of a deal if an extra rank is reported; I think it's more important to avoid missing URLs.

The ability to trigger it through a comment (`!AlexaBot! ...`) would still be crucial.

A preliminary blacklist:

```
github.com
githubusercontent.com
wikimedia.org
alexa.com
wikipedia.org
travis-ci.org
onlinewebfonts.com
rawgit.com
travis-ci.com
cdnjs.com
simpleicons.org
shareicon.net
brandcolors.net
imgur.com
pastebin.com
unpkg.com
github.io
google.com
seekvectorlogo.com
dropbox.com
```

<details><summary>Generated using: (code shouldn't be re-used in bot, as the domain extraction is shite)</summary>

```js
(function(){
/** (since: string) => Promise{issue[]} */
function loadIssues(since) {
  let url = "//api.github.com/repos/simple-icons/simple-icons/issues?sort=updated&direction=asc&state=all";
  if (since) { url += "&since="+since; }
  return fetch(url)
    .then(body => body.json());
}
/** (issues: issue[]) => string[] */
const URL_MATCHER = /\b(https?:(\/\/)?)?([^/.\s]+\.)+(?!(svg|png|json|md|js))[a-z]{2,5}\b/gi;
function getURLsFromIssue(issues) {
  const body = issues.body;
  return body.match(URL_MATCHER) || [];
}
/** (maxPages: number, since: string) => Promise{issue[]} */
async function loadAllIssues(maxPages = 10, since = "2018-01-01T00:00:00Z") {
  const seenIssues = {};
  let issues = [];
  let shouldRun = true;
  let i = 0;
  while (shouldRun) {
    ++i;
    const isss = await loadIssues(since);
    if (isss.length === 1) { break; }
    issues = issues.concat(isss.filter(issue => !(issue.number in seenIssues)));
    isss.forEach(issue => seenIssues[issue.number] = true);
    since = isss[isss.length - 1].updated_at;
    shouldRun = i < maxPages;
    await (new Promise(res => setTimeout(res, 500)));
  }
  return issues;
}

/** Turn all issues into URLs, count them up, log the result */
const seen = {};
const DOMAIN_MATCH = /[^\.\/]+\.[^\.]+$/;
loadAllIssues(99).then(issues => issues.reduce((prev,issue)=>prev.concat(getURLsFromIssue(issue)), []))
  .then(urls=>urls.forEach(url=>seen[url] = (seen[url] || 0)+1))
  .then(() => {
    console.log(Object.keys(seen)
      .sort((a,b) => seen[b]-seen[a])
      .map(url => DOMAIN_MATCH.exec(url)[0])
      .join("\n")
    );
  });
})();
```

</details>

**Comment by metaa** [comment]
What about [GitHub Checks](https://developer.github.com/v3/checks/) instead of comments?

This is an example mockup I made:
![image](https://user-images.githubusercontent.com/5056880/46021456-02167180-c0e1-11e8-95a3-ac9d5290f846.png)

**Comment by davidklebanoff** [comment]
@metaa That's a very cool idea and one I hadn't considered. Unfortunately, like you discovered, once we're at the point of a pull request, it's a bit late in the process.

Still, it may be worth looking into at some point to supplement whatever solution we settle on for issues. It may be worth putting in a separate issue.

**Comment by metaa** [comment]
We could define a template then and add a `shield` to it, asking the user to drop their url at the end of the image URL and then automatically assign an issue based on the Alexa rank.

##### [Badge](https://shields.io)
The badge (a bit chunky, but that's how it works):
```
https://img.shields.io/badge/dynamic/xml.svg?longCache=true&label=Alexa%20Rank&query=%2F%2FALEXA%2FSD%2FPOPULARITY%2F%40TEXT&colorB=blue&url=https%3A%2F%2Fdata.alexa.com%2Fdata%3Fcli%3D10%26url%3D<YOUR_URL>
```

For example:
[![Badge](https://img.shields.io/badge/dynamic/xml.svg?longCache=true&label=Alexa%20Rank&query=%2F%2FALEXA%2FSD%2FPOPULARITY%2F%40TEXT&colorB=blue&url=https%3A%2F%2Fdata.alexa.com%2Fdata%3Fcli%3D10%26url%3Dsimpleicons.org)](https://img.shields.io/badge/dynamic/xml.svg?longCache=true&label=Alexa%20Rank&query=%2F%2FALEXA%2FSD%2FPOPULARITY%2F%40TEXT&colorB=blue&url=https%3A%2F%2Fdata.alexa.com%2Fdata%3Fcli%3D10%26url%3Dsimpleicons.org)
(that's `simpleicons.org`)

**Comment by ericcornelissen** [comment]
I like both of your ideas @metaa. Even though GitHub Checks only works for Pull Request, I still think it is valuable to add it as it automatically blocks PRs for us, in that way we don't have to (potentially) read to a list of comments to find the rank.

The badge is a neat idea, however as you said the URL is a bit chunky. If we could somehow make it so issuers/contributors don't have to see the entire URL I'm all for it 👍

**Comment by davidklebanoff** [comment]
I currently have a prototype bot I've written using the Probot framework. I hope to publish the source in a separate simple-icons repository soon and eventually turn it on for this project.

It currently searches for links to official websites in new issues that at least marginally follow the "Icon Request" template, validates the link and checks it against the blacklist, calls an Alexa API to get the current ranking, and reports the rank in a follow-up comment.

It still needs some additional unit testing for edge cases and failure modes, but it's working well in a private repository I've been testing in.

**Comment by ericcornelissen** [comment]
Awesome! Can't wait to see the source and have it working in this repos :smile:

**Comment by ericcornelissen** [comment]
@davidklebanoff, may I ask how you're hosting the bot right now?

The service I'm using for the [svg-to-pdf-bot](https://github.com/simple-icons/svg-to-pdf-bot) is [ZEIT](https://zeit.co/), which was then recommended in the [Probot](https://github.com/probot/probot) documentation. And I just found out that it support organisations, which would make it easy to have all the bots in one place. Of course, we could also move that bot over to whatever you're using to host your bot :slightly_smiling_face:

**Comment by davidklebanoff** [comment]
Since it's under development, I just have it running locally with the [smee](https://smee.io) service routing the webhooks. I haven't put much thought into where to host the production version, but if ZEIT seems to be working well for you, then I'm happy to use it as well.

[GitHub Actions](https://github.com/features/actions) may be another option.

Either way, I think it would be wise to run both bots on the same host.

**Comment by ericcornelissen** [comment]
Agreed. Actually, I'm up for GitHub Actions as well In the long run that should be a better solution :smile: We just have to sign up the organisation for the Public beta before we can get started on that. I don't think we'll be doing anything that might not work in the beta [after 5 months](https://github.blog/2018-10-17-action-demos/) anyway.

**Comment by davidklebanoff** [comment]
I added Simple Icons to the beta list. We'll see if an invite somes anytime soon.

**Comment by ericcornelissen** [comment]
> I added Simple Icons to the beta list. We'll see if an invite comes anytime soon.

Since we still haven't heard anything(?), what do you think about letting the bot run on the repository without GitHub actions for now and see if it works and if we like it @davidklebanoff ?

**Comment by davidklebanoff** [comment]
I think that's fine. Unfortunately I'm quite busy for the foreseeable future with non-simple icons obligations and don't have time to dedicate to the bot at this time. I hope to revisit when things free up a bit.

**Comment by ericcornelissen** [comment]
> I think that's fine. Unfortunately I'm quite busy for the foreseeable future with non-simple icons obligations and don't have time to dedicate to the bot at this time. I hope to revisit when things free up a bit.

Care to share the source code of the bot with me so I can maybe look into getting it working on this repository?

**Comment by runxel** [comment]
> Care to share the source code of the bot with me so I can maybe look into getting it working on this repository?

Yeah, that would be awesome! 🌮

**Comment by mondeja** [comment]
We are not using Alexa rankings anymore and this issue has not activity in so long, so closing it...
